### PR TITLE
Allow batch size of 0 in Conv

### PR DESCRIPTION
**Pull request by jamesr66a**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#26214 [WIP] Allow batch size of 0 in Conv**

Differential Revision: [D17377035](https://our.internmc.facebook.com/intern/diff/D17377035)

This PR allows you to call convolution with a batch size of 0. In this case, we will simply return a properly-sized conv output (as if conv were performed), similarly with a batch size 0 and 0 elements.

NOTE: We are skipping the implementation for ROCm here, instead electing to make it a follow-up task.

### CODE REVIEW

**Review by fmassa — APPROVED** [review]
The approach looks good to me, thanks a lot @jamesr66a !

I would have hoped that we wouldn't need to special-case the double-backwards implementation to handle empty batch sizes because they rely under-the-hood on convolution, but it seems it's not possible and we need to handle it explicitly anyway?

**Comment by jamesr66a** [comment]
@fmassa yeah the double-backwards stuff is not great. What I found in practice is that the 0 dimensions in double-backward appear in places other than the batch dimension in some places, and that makes it so we have to special-case things. @zdevito has the impression that the double-backward implementation itself is more complicated than it needs to be. maybe it's something we can investigate later

File: aten/src/ATen/native/miopen/Conv_miopen.cpp [hunk 1]
```diff
@@ -669,6 +669,10 @@ Tensor miopen_convolution_forward(
                                      padding, stride, dilation, groups),
                     input->options());
 
+  if (output_t.numel() == 0) {
+    return output_t;
```
Comment (v0dro) [inline aten/src/ATen/native/miopen/Conv_miopen.cpp]: Did the necessity of making backend-level changes arise due to the implementation being scattered between TH and ATen? I'm working on `ConvTranspose` for 0-dim tensors and all it  is in ATen, which I think would be helpful in simply working with the `_convolution` function for the forward pass and leaving the rest untouched.

**Pull request closed without merging**